File: rl/chapter02/k_armed_bandit_testbed_test.cc
#include "rl/chapter02/k_armed_bandit_testbed.h"

#include "absl/log/log.h"
#include "gtest/gtest.h"

#include "rl/testing/random_seed.h"

namespace rl::test {

// A simple test to make sure that KArmedBanditTestbed works as intended.
TEST(KArmedBanditTestbedTest, TenArmedBandit) {
  KArmedBanditTestbed<10> testbed{testing::RandomSeed()};
  double reward{0.0};
  for (int i = 0; i < 100; ++i) {
    for (int j = 0; j < 10; ++j) {
      reward += testbed.PullArm(j);
    }
  }
  LOG(INFO) << "Total reward: " << reward;
}

} // namespace rl::test

File: rl/chapter02/k_armed_bandit_testbed_test.py
import numpy as np
import unittest
from rl.chapter02.k_armed_bandit_testbed import KArmedBanditTestbed


class KArmedBanditTestbedTest(unittest.TestCase):
    def setUp(self):
        self.testbed = KArmedBanditTestbed(10)

    def test_pull_arm_default(self):
        """Test that we can pull all arms and get the expected reward."""
        reward = 0.0
        for i in range(100):
            for j in range(self.testbed.k):
                reward += self.testbed.pull_arm(j)
        self.assertAlmostEqual(reward, 167.38037556944178)

    def test_pull_arm_custom_count(self):
        """Test that we can pull an arm `count` times with one call."""
        reward = self.testbed.pull_arm(0, 100)
        self.assertIsInstance(reward, np.ndarray)
        self.assertEqual(reward.shape, (100,))


if __name__ == "__main__":
    unittest.main()
